astropy/coordinates/tests/test_frames.py

```diff
@@ -283,6 +274,56 @@ class FakeICRS(ICRS):
     assert fi.ra.unit == u.hourangle
 
 
+def test_representation_info():
+    from ..baseframe import RepresentationMapping
+    from ..builtin_frames import ICRS
+
+    class NewICRS1(ICRS):
+        frame_specific_representation_info = {
+            r.SphericalRepresentation: [
+                RepresentationMapping('lon', 'rara', u.hourangle),
+                RepresentationMapping('lat', 'decdec', u.degree),
+                RepresentationMapping('distance', 'distance', u.kpc)]
+        }
+
+    i1 = NewICRS1(rara=10*u.degree, decdec=-12*u.deg, distance=1000*u.pc,
+                  pm_rara_cosdecdec=100*u.mas/u.yr,
+                  pm_decdec=17*u.mas/u.yr,
+                  radial_velocity=10*u.km/u.s)
+    assert allclose(i1.rara, 10*u.deg)
+    assert i1.rara.unit == u.hourangle
+    assert allclose(i1.decdec, -12*u.deg)
+    assert allclose(i1.distance, 1000*u.pc)
+    assert i1.distance.unit == u.kpc
+    assert allclose(i1.pm_rara_cosdecdec, 100*u.mas/u.yr)
+    assert allclose(i1.pm_decdec, 17*u.mas/u.yr)
+
+    # this should auto-set the names of UnitSpherical:
+    i1.set_representation_cls(r.UnitSphericalRepresentation,
+                              s=r.UnitSphericalCosLatDifferential)
+    assert allclose(i1.rara, 10*u.deg)
+    assert allclose(i1.decdec, -12*u.deg)
+    assert allclose(i1.pm_rara_cosdecdec, 100*u.mas/u.yr)
+    assert allclose(i1.pm_decdec, 17*u.mas/u.yr)
+
+    # For backwards compatibility, we also support the string name in the
+    # representation info dictionary:
+    class NewICRS2(ICRS):
+        frame_specific_representation_info = {
+            'spherical': [
+                RepresentationMapping('lon', 'ang1', u.hourangle),
+                RepresentationMapping('lat', 'ang2', u.degree),
+                RepresentationMapping('distance', 'howfar', u.kpc)]
+        }
+
+    i2 = NewICRS2(ang1=10*u.degree, ang2=-12*u.deg, howfar=1000*u.pc)
+    assert allclose(i2.ang1, 10*u.deg)
+    assert i2.ang1.unit == u.hourangle
+    assert allclose(i2.ang2, -12*u.deg)
+    assert allclose(i2.howfar, 1000*u.pc)
+    assert i2.howfar.unit == u.kpc
```

Comment:
Can you add a test here for just the radial differential? This may mean also defining another frame that sets `SphericalCosLatDifferential` explicitly.